### ISSUE TITLE
🐜 [BUG] Kagami crashes when using the secp256k1 hashing algorithm.

### ISSUE THREAD

**Issue by 6r1d** [hyperledger/iroha#2554]
While adding updates in https://github.com/hyperledger/iroha-2-docs/pull/92, specifically [this part](https://github.com/hyperledger/iroha-2-docs/pull/92#discussion_r931979388), I saw a bug in Kagami.

### GIT commit hash

`7ed7b4c80` and more

### Minimum working example

Kagami is using a [Secp256k1](https://en.bitcoin.it/wiki/Secp256k1) algorithm to generate a key pair. Both `cargo run` and executable show the same trace.

```bash
./target/debug/kagami crypto -s 86 -a secp256k1
```

```
cargo run --bin kagami  -- crypto -s 86 -a secp256k1
```

### Expected behaviour

I expect a key pair to be generated with `secp256k1` algorithm when a seed is provided, similar as to when it is disabled:

```bash
$ ./target/debug/kagami crypto -a secp256k1 
Public key (multihash): e701210250a46ea6e1688d1ac64fa653eb6220ea21591eb8cb850695c64b04d95853d527
Private key: d46c33e69e0bd21de8cb969eb395bdf0d86c6727d73db917436508e42963bc74
Digest function: secp256k1
```

### Actual behaviour

Kagami crashes in both modes with a trace provided below.

### Operating system

Artix Linux, kernel `5.18.12`

### LibC type and version

GNU libc v2.35

### Current environment

* Source code build, bare metal
* Source code build in Docker v20.10.17

### Trace

<details>
  <summary>Short trace</summary>
  
  ```text
The application panicked (crashed).
Message:  range end index 32 out of range for slice of length 2
Location: ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ursa-0.3.7/src/signatures/secp256k1.rs:152

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 7 frames hidden ⋮                               
   8: core::slice::index::slice_end_index_len_fail::h84caae524e731065
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/slice/index.rs:43
   9: <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[T]>>::index::h64b3a97fa59b684d
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/slice/index.rs:245
  10: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index::h99b48a44af85455a
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/slice/index.rs:15
  11: ursa::signatures::secp256k1::ecdsa_secp256k1::EcdsaSecp256k1Impl::keypair::ha0c7745b3ceac498
      at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ursa-0.3.7/src/signatures/secp256k1.rs:152
       150 │                     KeyGenOption::UseSeed(ref mut seed) => {
       151 │                         let mut s = [0u8; PRIVATE_KEY_SIZE];
       152 >                         let mut rng = ChaChaRng::from_seed(*array_ref!(seed.as_slice(), 0, 32));
       153 │                         seed.zeroize();
       154 │                         rng.fill_bytes(&mut s);
  12: <ursa::signatures::secp256k1::EcdsaSecp256k1Sha256 as ursa::signatures::SignatureScheme>::keypair::he3189a8382f41a75
      at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ursa-0.3.7/src/signatures/secp256k1.rs:33
        31 │         option: Option<KeyGenOption>,
        32 │     ) -> Result<(PublicKey, PrivateKey), CryptoError> {
        33 >         self.0.keypair::<sha2::Sha256>(option)
        34 │     }
        35 │     fn sign(&self, message: &[u8], sk: &PrivateKey) -> Result<Vec<u8>, CryptoError> {
  13: iroha_crypto::KeyPair::generate_with_configuration::h672762d406fc5823
      at ~/lts/crypto/src/lib.rs:293
  14: kagami::crypto::Args::key_pair::{{closure}}::h34f1b463f1e1634e
      at ~/lts/tools/kagami/src/main.rs:132
  15: core::option::Option<T>::map_or_else::h63fc220a2edaabb4
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/option.rs:999
  16: kagami::crypto::Args::key_pair::h86457b8851369e08
      at ~/lts/tools/kagami/src/main.rs:114
  17: <kagami::crypto::Args as kagami::RunArgs<T>>::run::h0f99b077cc0f78b2
      at ~/lts/tools/kagami/src/main.rs:97
  18: <kagami::Args as kagami::RunArgs<T>>::run::h25d2588233aac602
      at ~/lts/tools/kagami/src/main.rs:57
  19: kagami::main::he4a7a79b628d86a7
      at ~/lts/tools/kagami/src/main.rs:24
  20: core::ops::function::FnOnce::call_once::h481ddac26d0f4c7e
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:227
                                ⋮ 15 frames hidden ⋮ 
  ```
</details>

<details>
  <summary>Longer trace</summary>
  
  ```text
The application panicked (crashed).
Message:  range end index 32 out of range for slice of length 2
Location: ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ursa-0.3.7/src/signatures/secp256k1.rs:152

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   1: color_eyre::config::PanicHook::panic_report::h8744de1f3eb8d7eb
      at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/color-eyre-0.5.11/src/config.rs:982
       980 │ 
       981 │         let backtrace = if capture_bt {
       982 >             Some(backtrace::Backtrace::new())
       983 │         } else {
       984 │             None
   2: color_eyre::config::PanicHook::into_panic_hook::{{closure}}::hd9816d9f18ad966f
      at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/color-eyre-0.5.11/src/config.rs:961
       959 │     ) -> Box<dyn Fn(&std::panic::PanicInfo<'_>) + Send + Sync + 'static> {
       960 │         Box::new(move |panic_info| {
       961 >             eprintln!("{}", self.panic_report(panic_info));
       962 │         })
       963 │     }
   3: std::panicking::rust_panic_with_hook::h3b7380e99b825b63
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:702
   4: std::panicking::begin_panic_handler::{{closure}}::h8e849d0710154ce0
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:588
   5: std::sys_common::backtrace::__rust_end_short_backtrace::hedcdaddbd4c46cc5
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:138
   6: rust_begin_unwind<unknown>
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584
   7: core::panicking::panic_fmt::he1bbc7336d49a357
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143
   8: core::slice::index::slice_end_index_len_fail::h84caae524e731065
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/slice/index.rs:43
   9: <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[T]>>::index::h64b3a97fa59b684d
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/slice/index.rs:245
  10: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index::h99b48a44af85455a
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/slice/index.rs:15
  11: ursa::signatures::secp256k1::ecdsa_secp256k1::EcdsaSecp256k1Impl::keypair::ha0c7745b3ceac498
      at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ursa-0.3.7/src/signatures/secp256k1.rs:152
       150 │                     KeyGenOption::UseSeed(ref mut seed) => {
       151 │                         let mut s = [0u8; PRIVATE_KEY_SIZE];
       152 >                         let mut rng = ChaChaRng::from_seed(*array_ref!(seed.as_slice(), 0, 32));
       153 │                         seed.zeroize();
       154 │                         rng.fill_bytes(&mut s);
  12: <ursa::signatures::secp256k1::EcdsaSecp256k1Sha256 as ursa::signatures::SignatureScheme>::keypair::he3189a8382f41a75
      at ~/.cargo/registry/src/github.com-1ecc6299db9ec823/ursa-0.3.7/src/signatures/secp256k1.rs:33
        31 │         option: Option<KeyGenOption>,
        32 │     ) -> Result<(PublicKey, PrivateKey), CryptoError> {
        33 >         self.0.keypair::<sha2::Sha256>(option)
        34 │     }
        35 │     fn sign(&self, message: &[u8], sk: &PrivateKey) -> Result<Vec<u8>, CryptoError> {
  13: iroha_crypto::KeyPair::generate_with_configuration::h672762d406fc5823
      at ~/lts/crypto/src/lib.rs:293
  14: kagami::crypto::Args::key_pair::{{closure}}::h34f1b463f1e1634e
      at ~/lts/tools/kagami/src/main.rs:132
  15: core::option::Option<T>::map_or_else::h63fc220a2edaabb4
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/option.rs:999
  16: kagami::crypto::Args::key_pair::h86457b8851369e08
      at ~/lts/tools/kagami/src/main.rs:114
  17: <kagami::crypto::Args as kagami::RunArgs<T>>::run::h0f99b077cc0f78b2
      at ~/lts/tools/kagami/src/main.rs:97
  18: <kagami::Args as kagami::RunArgs<T>>::run::h25d2588233aac602
      at ~/lts/tools/kagami/src/main.rs:57
  19: kagami::main::he4a7a79b628d86a7
      at ~/lts/tools/kagami/src/main.rs:24
  20: core::ops::function::FnOnce::call_once::h481ddac26d0f4c7e
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:227
  21: std::sys_common::backtrace::__rust_begin_short_backtrace::h26923d3df08fe806
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:122
  22: std::rt::lang_start::{{closure}}::h20181462de5043ac
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/rt.rs:145
  23: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hb7014f43484a8b4e
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:259
  24: std::panicking::try::do_call::h7bc9dc436daeb8c7
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:492
  25: std::panicking::try::h653d68a27ff5f175
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:456
  26: std::panic::catch_unwind::h9d739f9f59895e68
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panic.rs:137
  27: std::rt::lang_start_internal::{{closure}}::hf006f2bc7ce22bbe
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/rt.rs:128
  28: std::panicking::try::do_call::hfb39d6df61a2e69f
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:492
  29: std::panicking::try::h13e2d225134958ac
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:456
  30: std::panic::catch_unwind::h3bd49b5a5dfb1a50
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panic.rs:137
  31: std::rt::lang_start_internal::h2ba92edce36c035e
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/rt.rs:128
  32: std::rt::lang_start::hcfb778be6e0a78f0
      at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/rt.rs:144
  33: main<unknown>
      at <unknown source file>:<unknown line>
  34: __libc_start_main<unknown>
      at <unknown source file>:<unknown line>
  35: _start<unknown>
      at <unknown source file>:<unknown line>
  ```
</details>

### Who can help?

@s8sato @appetrosyan

**Comment by Erigara** [comment]
Problem is following `ursa` expect seed size of 32 bytes in case of  `secp256k1` and provided seed `82` is only is only 2 bytes long.
The following example working: 
```
cargo run --bin kagami  -- crypto -s 00000000000000000000000000000000 -a secp256k1
```
However for `ed25519` everything seems to work fine.

Difference is the following:
1. `ed25519` compute `sha256` hash of received function and than takes 32 bytes from hash to feed rng generator.
2. `secp256k1` directly feed input string to rng generator, which expect to recieve at least 32bytes long input.
I'm not sure if this is a bug or intended behavior, @appetrosyan what do you think? 

Basically i think there is 2 workaround solution for this:
1. Hash on our size and provide seed of proper size to ursa (one downside to this is that user will recieve different result from previous version of kagami with the same seed).
2. Rise error with meaningful message to user (not really like this as this would require user to provide 32 character long strings as seed).
